File: app/src/main/java/com/bkjcb/rqapplication/check/CheckResultDetailActivity.java
package com.bkjcb.rqapplication.check;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;

import com.bkjcb.rqapplication.Constants;
import com.bkjcb.rqapplication.MyApplication;
import com.bkjcb.rqapplication.R;
import com.bkjcb.rqapplication.SimpleBaseActivity;
import com.bkjcb.rqapplication.check.fragment.ApplianceCheckResultFragment;
import com.bkjcb.rqapplication.check.fragment.CheckResultFragment;
import com.bkjcb.rqapplication.check.model.CheckItem;
import com.bkjcb.rqapplication.check.model.CheckResultItem;
import com.bkjcb.rqapplication.check.model.CheckResultItem_;
import com.bkjcb.rqapplication.check.retrofit.CheckService;
import com.bkjcb.rqapplication.ftp.FtpUtils;
import com.bkjcb.rqapplication.ftp.UploadTask;
import com.bkjcb.rqapplication.model.ExportFilePathResult;
import com.bkjcb.rqapplication.model.HttpResult;
import com.bkjcb.rqapplication.retrofit.NetworkApi;
import com.bkjcb.rqapplication.util.Utils;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import butterknife.BindView;
import butterknife.OnClick;
import io.reactivex.ObservableSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by DengShuai on 2019/12/19.
 * Description :
 */
public class CheckResultDetailActivity extends SimpleBaseActivity {

    @BindView(R.id.info_operation)
    Button mInfoOperation;
    @BindView(R.id.info_export)
    Button mInfoExport;
    protected CheckItem checkItem;
    private CheckResultFragment fragment;

    @Override
    protected int setLayoutID() {
        return R.layout.activity_detail_result_check;
    }

    @Override
    protected void initView() {
        initTopbar("检查信息", 0);
    }

    @Override
    protected void initData() {
        checkItem = (CheckItem) getIntent().getSerializableExtra("data");
        if (checkItem == null) {
            long id = getIntent().getLongExtra("id", 0);
            checkItem = CheckItem.getBox().get(id);
            if (checkItem == null) {
                return;
            }
        }
        if (checkItem.type == 0) {
            fragment = CheckResultFragment.newInstance(checkItem);
        } else {
            fragment = ApplianceCheckResultFragment.newInstance(checkItem);
        }
        showCheckDetail();
        //initTextValue();
        mInfoExport.setVisibility(checkItem.status == 3 ? View.VISIBLE : View.GONE);
        mInfoOperation.setText(getOperation(checkItem.status));
    }

    private void showCheckDetail() {
        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.content, fragment)
                .commit();
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        initTextValue();
        fragment.updateData(checkItem);
    }

    private void initTextValue() {
        checkItem = CheckItem.getBox().get(checkItem.id);
        mInfoOperation.setText(getOperation(checkItem.status));
    }

    @OnClick({R.id.info_operation, R.id.info_export})
    public void onClick(View v) {
        if (v.getId() == R.id.info_operation) {
            if (checkItem.status == 2) {
                submitResult();
            } else {
                toCheckContentActivity();
            }
        } else if (v.getId() == R.id.info_export) {
            getExportFilePath();
        }

    }

    protected void toCheckContentActivity() {
        CheckDetailActivity.ToActivity(this, checkItem.id);
    }

    protected static void ToActivity(Context context, CheckItem checkItem) {
        Intent intent = new Intent(context, CheckResultDetailActivity.class);
        intent.putExtra("data", checkItem);
        context.startActivity(intent);
    }

    protected static void ToActivity(Context context, long id) {
        Intent intent = new Intent(context, CheckResultDetailActivity.class);
        intent.putExtra("id", id);
        context.startActivity(intent);
    }

    private String getOperation(int status) {
        String retString = "";
        switch (status) {
            case 0:
                retString = "开始检查";
                break;
            case 1:
                retString = "继续检查";
                break;
            case 2:
                retString = "上传提交";
                break;
            case 3:
                retString = "查看详情";
            default:
        }
        return retString;
    }

    private List<String> getFiles() {
        if (TextUtils.isEmpty(checkItem.filePath)) {
            return new ArrayList<>();
        }
        return Arrays.asList(checkItem.filePath.split(","));
    }

    protected void submitResult() {
        showLoading(true);
        List<CheckResultItem> list = queryResult();
        disposable = UploadTask.createUploadTask(getFiles(), Utils.getFTPPath(checkItem), new FtpUtils.UploadProgressListener() {
            @Override
            public void onUploadProgress(String currentStep, long uploadSize, long size, File file) {

            }
        }).subscribeOn(Schedulers.io())
                .flatMap(new Function<Boolean, ObservableSource<HttpResult>>() {
                    @Override
                    public ObservableSource<HttpResult> apply(Boolean aBoolean) throws Exception {
                        return aBoolean ? NetworkApi.getService(CheckService.class).saveCheckItem(
                                MyApplication.getUser().getUserId(),
                                checkItem.year,
                                checkItem.zhandianleixing,
                                checkItem.beijiandanweiid,
                                checkItem.beizhu,
                                checkItem.jianchariqi,
                                checkItem.jianchajieguo,
                                getItemsID(list),
                                getItemsResult(list),
                                Utils.getFTPPath(checkItem),
                                checkItem.c_id
                        ) : null;
                    }
                })
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<HttpResult>() {
                    @Override
                    public void accept(HttpResult result) throws Exception {
                        if (result.pushState == 200) {
                            updateTaskStatus();
                        }
                        showTipInfo(result.pushMsg);
                        showLoading(false);
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        showLoading(false);
                        showTipInfo("上传失败！" + throwable.getMessage());
                    }
                });
    }

    protected void updateTaskStatus() {
        checkItem.status = 3;
        CheckItem.getBox().put(checkItem);
        initTextValue();
    }

    protected void showTipInfo(String info) {
        showSnackbar(mInfoOperation, info);
    }

    private String[] getItemsID(List<CheckResultItem> list) {
        String[] strings = new String[list.size()];
        for (int i = 0; i < list.size(); i++) {
            strings[i] = list.get(i).jianchaxiangid;
        }
        return strings;
    }

    private String[] getItemsResult(List<CheckResultItem> list) {
        String[] strings = new String[list.size()];
        for (int i = 0; i < list.size(); i++) {
            strings[i] = TextUtils.isEmpty(list.get(i).jianchajilu) ? "" : list.get(i).jianchajilu;
        }
        return strings;
    }

    private List<CheckResultItem> queryResult() {
        return CheckResultItem.getBox().query().equal(CheckResultItem_.jianchaid, checkItem.c_id).build().find();
    }

    private void getExportFilePath() {
        showLoading(true);
        disposable = NetworkApi.getService(CheckService.class)
                .getExportPath(checkItem.c_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<ExportFilePathResult>() {
                    @Override
                    public void accept(ExportFilePathResult exportFilePathResult) throws Exception {
                        showLoading(false);
                        if (exportFilePathResult.pushState == 200) {
                            Intent intent = new Intent(Intent.ACTION_VIEW);
                            intent.setData(Uri.parse(Constants.BASE_URL + "/rq/push/downloadfile?downloadfile=" + exportFilePathResult.getDatas()));
                            startActivity(intent);
                        } else {
                            showSnackbar(mInfoExport, "生成文件失败！" + exportFilePathResult.getPushMsg());
                        }
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        showLoading(false);
                        showSnackbar(mInfoExport, "获取文件地址失败！" + throwable.getMessage());
                    }
                });
    }
}

File: app/src/main/java/com/bkjcb/rqapplication/check/CheckResultDetailActivity.kt
package com.bkjcb.rqapplication.check

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.text.TextUtils
import android.view.View
import com.bkjcb.rqapplication.base.BaseSimpleActivity
import com.bkjcb.rqapplication.Constants
import com.bkjcb.rqapplication.MyApplication
import com.bkjcb.rqapplication.R
import com.bkjcb.rqapplication.check.fragment.ApplianceCheckResultFragment
import com.bkjcb.rqapplication.check.fragment.CheckResultFragment
import com.bkjcb.rqapplication.check.model.CheckItem
import com.bkjcb.rqapplication.check.model.CheckResultItem
import com.bkjcb.rqapplication.check.model.CheckResultItem_
import com.bkjcb.rqapplication.check.retrofit.CheckService
import com.bkjcb.rqapplication.base.datebase.DataUtil
import com.bkjcb.rqapplication.base.ftp.FtpUtils.UploadProgressListener
import com.bkjcb.rqapplication.base.ftp.UploadTask.createUploadTask
import com.bkjcb.rqapplication.base.retrofit.NetworkApi.Companion.getService
import com.bkjcb.rqapplication.base.util.Utils.getFTPPath
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_detail_result_check.*
import kotlinx.android.synthetic.main.with_bg_top_bar.*
import java.util.*

/**
 * Created by DengShuai on 2019/12/19.
 * Description :
 */
open class CheckResultDetailActivity : BaseSimpleActivity(), View.OnClickListener {
    protected lateinit var checkItem: CheckItem
    private lateinit var fragment: CheckResultFragment
    override fun setLayoutID(): Int {
        return R.layout.activity_detail_result_check
    }

    override fun initView() {
        initTopBar("检查信息", appbar)
        info_operation.setOnClickListener(this)
        info_export.setOnClickListener(this)
    }

    override fun initData() {
        val item: CheckItem? = intent.getSerializableExtra("data") as CheckItem
        checkItem = if (item == null) {
            val id = intent.getLongExtra("id", 0)
            DataUtil.getCheckItemBox().get(id)
        } else {
            item
        }
        fragment = if (checkItem.type == 0) {
            CheckResultFragment.newInstance(checkItem)
        } else {
            ApplianceCheckResultFragment.newInstance(checkItem)
        }
        showCheckDetail()
        //initTextValue();
        info_export.visibility = if (checkItem.status == 3) View.VISIBLE else View.GONE
        info_operation.text = getOperation(checkItem.status)
    }

    private fun showCheckDetail() {
        supportFragmentManager
                .beginTransaction()
                .add(R.id.content, fragment)
                .commit()
    }

    override fun onRestart() {
        super.onRestart()
        initTextValue()
        fragment.updateData(checkItem)
    }

    private fun initTextValue() {
        checkItem = DataUtil.getCheckItemBox().get(checkItem.id)
        info_operation.text = getOperation(checkItem.status)
    }

    override fun onClick(v: View) {
        if (v.id == R.id.info_operation) {
            if (checkItem.status == 2) {
                submitResult()
            } else {
                toCheckContentActivity()
            }
        } else if (v.id == R.id.info_export) {
            initExportFilePath()
        }
    }

    protected open fun toCheckContentActivity() {
        CheckDetailActivity.toActivity(this, checkItem.id)
    }

    private fun getOperation(status: Int): String {
        return when (status) {
            0 -> "开始检查"
            1 -> "继续检查"
            2 -> "上传提交"
            3 -> "查看详情"
            else -> ""
        }
    }

    private fun initFiles(): List<String?> = if (TextUtils.isEmpty(checkItem.filePath)) {
        ArrayList()
    } else checkItem.filePath!!.split(",")

    protected open fun submitResult() {
        showLoading(true)
        val list = queryResult()
        disposable = createUploadTask(initFiles(), getFTPPath(checkItem), UploadProgressListener { currentStep, uploadSize, size, file -> }).subscribeOn(Schedulers.io())
                .flatMap { aBoolean ->
                    if (aBoolean) getService(CheckService::class.java).saveCheckItem(
                            MyApplication.getUser().userId,
                            checkItem.year,
                            checkItem.zhandianleixing,
                            checkItem.beijiandanweiid,
                            checkItem.beizhu,
                            checkItem.jianchariqi,
                            checkItem.jianchajieguo,
                            getItemsID(list),
                            getItemsResult(list),
                            getFTPPath(checkItem),
                            checkItem.c_id
                    ) else null
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ result ->
                    if (result.pushState == 200) {
                        updateTaskStatus()
                    }
                    showTipInfo(result.pushMsg)
                    showLoading(false)
                }) { throwable ->
                    showLoading(false)
                    showTipInfo("上传失败！" + throwable.message)
                }
    }

    protected fun updateTaskStatus() {
        checkItem.status = 3
        DataUtil.getCheckItemBox().put(checkItem)
        initTextValue()
    }

    protected fun showTipInfo(info: String?) {
        showSnackbar(info_operation, info!!)
    }

    private fun getItemsID(list: List<CheckResultItem>): Array<String?> {
        val strings = arrayOfNulls<String>(list.size)
        for (i in list.indices) {
            strings[i] = list[i].jianchaxiangid
        }
        return strings
    }

    private fun getItemsResult(list: List<CheckResultItem>): Array<String?> {
        val strings = arrayOfNulls<String>(list.size)
        for (i in list.indices) {
            strings[i] = if (TextUtils.isEmpty(list[i].jianchajilu)) "" else list[i].jianchajilu
        }
        return strings
    }

    private fun queryResult(): List<CheckResultItem> {
        return DataUtil.getCheckResultItemBox().query().equal(CheckResultItem_.jianchaid, checkItem.c_id!!).build().find()
    }

    private fun initExportFilePath() {
        showLoading(true)
        disposable = getService(CheckService::class.java)
                .getExportPath(checkItem.c_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ exportFilePathResult ->
                    showLoading(false)
                    if (exportFilePathResult.pushState == 200) {
                        val intent = Intent(Intent.ACTION_VIEW)
                        intent.data = Uri.parse(Constants.BASE_URL + "/rq/push/downloadfile?downloadfile=" + exportFilePathResult.datas)
                        startActivity(intent)
                    } else {
                        showSnackbar(info_export, "生成文件失败！" + exportFilePathResult.pushMsg)
                    }
                }) { throwable ->
                    showLoading(false)
                    showSnackbar(info_export, "获取文件地址失败！" + throwable.message)
                }
    }

    companion object {
        fun toActivity(context: Context, checkItem: CheckItem?) {
            val intent = Intent(context, CheckResultDetailActivity::class.java)
            intent.putExtra("data", checkItem)
            context.startActivity(intent)
        }

        fun toActivity(context: Context, id: Long) {
            val intent = Intent(context, CheckResultDetailActivity::class.java)
            intent.putExtra("id", id)
            context.startActivity(intent)
        }
    }
}